File: wots/sync.py
"""
Sync operations: stow (user/root) + copy-sync (winuser/mnt) + async batch.

All Windows operations use **copy** (not symlink) for maximum compatibility.
File comparison is based on mtime + size.
"""
from __future__ import annotations

import asyncio
import shutil
import subprocess
from pathlib import Path
from typing import Callable, Dict, List, Optional, Tuple

from . import config
from .types import PkgType
from .discover import list_syncable_files, winuser_rel_path
from .utils import has_stow, has_pwsh, is_wsl, is_excluded


# ═══════════════════════════════════════════════════════════════════════════════
#  Stow (user / root)
# ═══════════════════════════════════════════════════════════════════════════════

def do_stow(
    pkg: Path,
    target: Path,
    *,
    sudo: bool = False,
    dry_run: bool = False,
) -> bool:
    """Run `stow --adopt -t <target> <pkg>` inside DOTFILES_DIR."""
    from .display import success, error, dim
    from .utils import run

    if not has_stow():
        error("GNU Stow is not installed — cannot stow packages.")
        return False

    pkg_name = pkg.name
    try:
        run(
            ["stow", "-v", "--adopt", "-t", str(target), pkg_name],
            sudo=sudo, cwd=config.DOTFILES_DIR, dry_run=dry_run,
        )
        if dry_run:
            dim(f"    DRY-RUN  would stow  {pkg_name}  →  {target}")
        else:
            success(f"Stowed  {pkg_name}  →  {target}")
        return True
    except subprocess.CalledProcessError:
        return False


# ═══════════════════════════════════════════════════════════════════════════════
#  Windows copy (sync, blocking)
# ═══════════════════════════════════════════════════════════════════════════════

def _win_copy(
    wsl_src: Path,
    win_dst: str,       # e.g. "C:\\Users\\zion\\.gitconfig"
    *,
    is_dir: bool = False,
    dry_run: bool = False,
) -> bool:
    """
    Copy a file / directory from WSL to Windows via  pwsh.exe + cmd /c copy|xcopy.
    """
    from .display import dim, error

    wsl_unc = f"\\\\wsl$\\{config.WSL_DISTRO_NAME}" + str(wsl_src).replace("/", "\\")

    if is_dir:
        parts = ["xcopy", "/E", "/I", "/Y", f'"{wsl_unc}"', f'"{win_dst}"']
    else:
        parts = ["copy", "/Y", f'"{wsl_unc}"', f'"{win_dst}"']

    cmd_str = " ".join(parts)
    full = ["pwsh.exe", "-NoProfile", "-Command", f"cmd /c {cmd_str}"]
    dim(f"  WIN  {' '.join(full)}")

    if dry_run:
        return True

    try:
        r = subprocess.run(
            full, capture_output=True, text=True,
            cwd=str(config.MNT_C), encoding="gbk", errors="replace",
        )
        if r.stdout.strip():
            dim(f"       {r.stdout.strip()}")
        if r.stderr.strip():
            dim(f"       {r.stderr.strip()}")
        if r.returncode != 0:
            error(f"Windows copy failed (rc={r.returncode}): {cmd_str}")
            return False
        return True
    except Exception as e:
        error(f"Exception running Windows copy: {e}")
        return False


def _wsl_copy(
    win_src: Path,        # /mnt/c/Users/zion/...
    wsl_dst: Path,        # c.mnt/Users/zion/...  or  {pkg}.winuser/zion/...
    *,
    is_dir: bool = False,
    dry_run: bool = False,
) -> bool:
    """Copy a file / directory from Windows (/mnt/c) into WSL."""
    from .display import dim, error

    if dry_run:
        dim(f"    DRY-RUN  would copy  {win_src}  →  {wsl_dst}")
        return True

    try:
        wsl_dst.parent.mkdir(parents=True, exist_ok=True)
        if is_dir:
            shutil.copytree(str(win_src), str(wsl_dst), dirs_exist_ok=True)
        else:
            shutil.copy2(str(win_src), str(wsl_dst))
        return True
    except Exception as e:
        error(f"Copy failed: {e}")
        return False


# ═══════════════════════════════════════════════════════════════════════════════
#  Async Windows copy (for batch operations)
# ═══════════════════════════════════════════════════════════════════════════════

async def _win_copy_async(
    wsl_src: Path,
    win_dst: str,
    *,
    is_dir: bool = False,
    dry_run: bool = False,
) -> bool:
    """Async version of _win_copy — uses asyncio subprocess."""
    from .display import dim, error

    wsl_unc = f"\\\\wsl$\\{config.WSL_DISTRO_NAME}" + str(wsl_src).replace("/", "\\")
    if is_dir:
        parts = ["xcopy", "/E", "/I", "/Y", f'"{wsl_unc}"', f'"{win_dst}"']
    else:
        parts = ["copy", "/Y", f'"{wsl_unc}"', f'"{win_dst}"']

    cmd_str = " ".join(parts)
    full = ["pwsh.exe", "-NoProfile", "-Command", f"cmd /c {cmd_str}"]

    if dry_run:
        return True

    try:
        proc = await asyncio.create_subprocess_exec(
            *full,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            cwd=str(config.MNT_C),
        )
        stdout, stderr = await proc.communicate()
        if proc.returncode != 0:
            error(f"Windows copy failed (rc={proc.returncode}): {cmd_str}")
            if stderr:
                dim(f"       {stderr.decode('gbk', errors='replace')}")
            return False
        return True
    except Exception as e:
        error(f"Exception: {e}")
        return False


async def _wsl_copy_async(
    win_src: Path,
    wsl_dst: Path,
    *,
    is_dir: bool = False,
    dry_run: bool = False,
) -> bool:
    """Async WSL-side copy (runs in thread to avoid blocking)."""
    loop = asyncio.get_running_loop()
    return await loop.run_in_executor(
        None, _wsl_copy, win_src, wsl_dst, is_dir, dry_run,
    )


# ═══════════════════════════════════════════════════════════════════════════════
#  Single file sync (sync version — used by create)
# ═══════════════════════════════════════════════════════════════════════════════

def sync_one_file(
    wsl_path: Path,
    win_path: Path,
    *,
    direction: str = "sync",
    dry_run: bool = False,
) -> str:
    """
    Sync a single file between WSL and Windows.

    direction:
      "to-windows"  — WSL → Windows (push)
      "to-wsl"      — Windows → WSL (pull)
      "sync"        — bidirectional, newer wins

    Returns one of:
      'copied_to_win', 'copied_to_wsl', 'up_to_date',
      'missing_source', 'error'
    """
    from .display import dim, warning, error

    wsl_exists = wsl_path.exists()
    try:
        win_exists = win_path.exists() or win_path.is_symlink()
    except OSError:
        win_exists = True

    # ── direction: to-windows ──
    if direction == "to-windows":
        if not wsl_exists:
            warning(f"WSL source missing: {wsl_path}")
            return "missing_source"
        if _win_copy(wsl_path, _win_path_str(win_path),
                     is_dir=wsl_path.is_dir(), dry_run=dry_run):
            return "copied_to_win"
        return "error"

    # ── direction: to-wsl ──
    if direction == "to-wsl":
        if not win_exists:
            warning(f"Windows source missing: {win_path}")
            return "missing_source"
        if _wsl_copy(win_path, wsl_path, is_dir=win_path.is_dir(), dry_run=dry_run):
            return "copied_to_wsl"
        return "error"

    # ── direction: sync (bidirectional) ──
    if wsl_exists and not win_exists:
        if _win_copy(wsl_path, _win_path_str(win_path),
                     is_dir=wsl_path.is_dir(), dry_run=dry_run):
            return "copied_to_win"
        return "error"

    if win_exists and not wsl_exists:
        if _wsl_copy(win_path, wsl_path, is_dir=win_path.is_dir(), dry_run=dry_run):
            return "copied_to_wsl"
        return "error"

    # both exist — compare
    try:
        ws = wsl_path.stat()
        wn = win_path.stat()
    except OSError as e:
        error(f"Stat failed: {e}")
        return "error"

    if abs(ws.st_mtime - wn.st_mtime) < 1.0 and ws.st_size == wn.st_size:
        return "up_to_date"

    if ws.st_mtime > wn.st_mtime:
        if _win_copy(wsl_path, _win_path_str(win_path),
                     is_dir=wsl_path.is_dir(), dry_run=dry_run):
            return "copied_to_win"
        return "error"
    elif wn.st_mtime > ws.st_mtime:
        if _wsl_copy(win_path, wsl_path, is_dir=win_path.is_dir(), dry_run=dry_run):
            return "copied_to_wsl"
        return "error"
    else:
        return "up_to_date"


# ═══════════════════════════════════════════════════════════════════════════════
#  Async single file sync (for batch)
# ═══════════════════════════════════════════════════════════════════════════════

async def _sync_one_async(
    wsl_path: Path,
    win_path: Path,
    *,
    direction: str = "sync",
    dry_run: bool = False,
    sem: asyncio.Semaphore | None = None,
) -> str:
    """Async version of sync_one_file — uses semaphore for concurrency."""
    from .display import dim, warning, error

    async def _do_win_copy():
        async with sem if sem else _null_context():
            return await _win_copy_async(
                wsl_path, _win_path_str(win_path),
                is_dir=wsl_path.is_dir(), dry_run=dry_run,
            )

    async def _do_wsl_copy():
        async with sem if sem else _null_context():
            return await _wsl_copy_async(
                win_path, wsl_path,
                is_dir=win_path.is_dir(), dry_run=dry_run,
            )

    wsl_exists = wsl_path.exists()
    try:
        win_exists = win_path.exists() or win_path.is_symlink()
    except OSError:
        win_exists = True

    if direction == "to-windows":
        if not wsl_exists:
            return "missing_source"
        ok = await _do_win_copy()
        return "copied_to_win" if ok else "error"

    if direction == "to-wsl":
        if not win_exists:
            return "missing_source"
        ok = await _do_wsl_copy()
        return "copied_to_wsl" if ok else "error"

    # sync
    if wsl_exists and not win_exists:
        ok = await _do_win_copy()
        return "copied_to_win" if ok else "error"
    if win_exists and not wsl_exists:
        ok = await _do_wsl_copy()
        return "copied_to_wsl" if ok else "error"

    try:
        ws = wsl_path.stat()
        wn = win_path.stat()
    except OSError:
        return "error"

    if abs(ws.st_mtime - wn.st_mtime) < 1.0 and ws.st_size == wn.st_size:
        return "up_to_date"
    if ws.st_mtime > wn.st_mtime:
        ok = await _do_win_copy()
        return "copied_to_win" if ok else "error"
    elif wn.st_mtime > ws.st_mtime:
        ok = await _do_wsl_copy()
        return "copied_to_wsl" if ok else "error"
    return "up_to_date"


# ═══════════════════════════════════════════════════════════════════════════════
#  Async batch sync (with progress)
# ═══════════════════════════════════════════════════════════════════════════════

def sync_batch(
    items: List[Tuple[Path, Path]],   # [(wsl_path, win_path), ...]
    *,
    direction: str = "sync",
    dry_run: bool = False,
    max_concurrent: int = 0,
    progress_cb: Callable[[str, int, int], None] | None = None,
) -> Dict[str, int]:
    """
    Sync a batch of (wsl_path, win_path) pairs in parallel.

    Returns counts by result category.
    """
    return asyncio.run(_sync_batch_async(
        items, direction=direction, dry_run=dry_run,
        max_concurrent=max_concurrent, progress_cb=progress_cb,
    ))


async def _sync_batch_async(
    items: List[Tuple[Path, Path]],
    *,
    direction: str = "sync",
    dry_run: bool = False,
    max_concurrent: int = 0,
    progress_cb: Callable[[str, int, int], None] | None = None,
) -> Dict[str, int]:
    """
    Async core of sync_batch.

    Uses a semaphore to cap concurrent Windows pwsh.exe calls.
    """
    from .display import dim

    mc = max_concurrent or config.MNT_MAX_CONCURRENT
    sem = asyncio.Semaphore(max(1, mc))

    counts: Dict[str, int] = {
        "copied_to_win": 0, "copied_to_wsl": 0, "up_to_date": 0,
        "missing_source": 0, "error": 0, "skipped": 0,
    }
    total = len(items)
    done = 0

    async def _one(wsl_p: Path, win_p: Path):
        nonlocal done
        # Skip excluded or too-large files
        if is_excluded(wsl_p):
            return "skipped"
        try:
            if config.MAX_SYNC_SIZE_BYTES > 0 and wsl_p.stat().st_size > config.MAX_SYNC_SIZE_BYTES:
                return "skipped"
        except OSError:
            return "error"

        result = await _sync_one_async(
            wsl_p, win_p, direction=direction, dry_run=dry_run, sem=sem,
        )
        done += 1
        if progress_cb:
            progress_cb(result, done, total)
        return result

    tasks = [_one(ws, wn) for ws, wn in items]
    results = await asyncio.gather(*tasks, return_exceptions=True)

    for r in results:
        if isinstance(r, Exception):
            counts["error"] += 1
        else:
            counts[r] = counts.get(r, 0) + 1

    return counts


# ═══════════════════════════════════════════════════════════════════════════════
#  Helpers
# ═══════════════════════════════════════════════════════════════════════════════

def _win_path_str(p: Path) -> str:
    """Convert a /mnt/c-based Path to a C:\\ string for Windows commands."""
    s = str(Path("C:/") / p.relative_to(config.MNT_C))
    return s.replace("/", "\\")


class _null_context:
    """Async no-op context manager."""
    async def __aenter__(self):
        return None
    async def __aexit__(self, *args):
        pass


def prepare_sync_items(
    pkg: Path, pt: PkgType,
) -> List[Tuple[Path, Path]]:
    """
    Build a list of (wsl_path, win_path) pairs for every syncable file
    in *pkg* (any copy-sync type).
    """
    from .discover import list_syncable_files, build_mnt_path
    items: List[Tuple[Path, Path]] = []

    if not pt.uses_copy_sync:
        return items

    files = list_syncable_files(pkg, pt)
    for f in files:
        win_path = build_mnt_path(f, pt)
        items.append((f, win_path))

    return items


def build_winuser_wsl_path(pkg: Path, win_path_on_mnt: Path, pt: PkgType | None = None) -> Path:
    """
    Given a package and a /mnt/c/... path, return the corresponding
    path inside the package (reverse of build_mnt_path).

    For winuser packages, the package root maps to C:\\Users\\{WIN_USERNAME}\\
    For winconfig,          → C:\\Users\\{WIN_USERNAME}\\.config\\
    etc.
    """
    # Try to figure out the relative path from the target
    target = pt.sync_target if pt else None
    if target:
        try:
            rel = win_path_on_mnt.relative_to(
                Path(str(target).replace("C:", str(config.MNT_C)))
            )
            return pkg / rel
        except ValueError:
            pass

    # Fallback: relative to /mnt/c
    try:
        rel = win_path_on_mnt.relative_to(config.MNT_C)
    except ValueError:
        rel = win_path_on_mnt
    return pkg / rel

File: wots/src/sync.rs
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::sync::atomic::{AtomicUsize, Ordering};
use std::sync::Arc;

use anyhow::{bail, Context, Result};
use rayon::prelude::*;

use crate::cli::SyncArgs;
use crate::config::{DOTFILES_DIR, MNT_C, SYNC_MAX_CONCURRENT, WSL_DISTRO_NAME};
use crate::discover::{build_win_path, find_packages, list_syncable_files, pkg_basename};
use crate::display;
use crate::types::{type_label, PkgType, SYNCABLE_TYPES};
use crate::util::{has_pwsh, has_robocopy, has_stow, is_excluded, is_wsl};

pub fn run(args: SyncArgs) -> Result<()> {
    let packages = find_packages(&DOTFILES_DIR);
    let types_to_sync: Vec<PkgType> = if let Some(pt) = &args.pkg_type {
        vec![*pt]
    } else {
        SYNCABLE_TYPES.to_vec()
    };

    let has_root = types_to_sync.contains(&PkgType::Root);
    if has_root && !args.bypass && !args.dry_run {
        display::info("\n⚠ Root sync requires sudo and will modify system files.");
        if !display::prompt::confirm("  Continue?", false) {
            display::info("Cancelled.");
            return Ok(());
        }
    }

    for pt in &types_to_sync {
        let pkgs = packages.get(pt).cloned().unwrap_or_default();
        let pkgs: Vec<PathBuf> = if let Some(ref app) = args.app {
            pkgs
                .into_iter()
                .filter(|p| pkg_basename(p) == *app)
                .collect()
        } else {
            pkgs
        };

        if pkgs.is_empty() {
            continue;
        }

        display::rule(&format!("Syncing {} packages", pt.value()));

        if pt.uses_stow() {
            sync_stow_batch(&pkgs, pt, args.dry_run, args.quiet)?;
        } else if pt.uses_copy_sync() {
            sync_copy_batch(&pkgs, pt, args.dry_run, args.quiet)?;
        } else {
            if !args.quiet {
                display::info("  (meta packages are manually managed)");
            }
        }
    }

    display::info("");
    display::success("Sync complete.");
    Ok(())
}

fn sync_stow_batch(pkgs: &[PathBuf], pt: &PkgType, dry_run: bool, quiet: bool) -> Result<()> {
    if !has_stow() {
        display::error("GNU Stow not installed — skipping.");
        return Ok(());
    }

    for pkg in pkgs {
        if !quiet {
            display::info(&format!(
                "  {}  →  {}",
                pkg.file_name()
                    .unwrap_or(std::ffi::OsStr::new(""))
                    .to_string_lossy(),
                type_label(*pt),
            ));
        }
        do_stow(pkg, pt, dry_run)?;
    }

    Ok(())
}

fn sync_copy_batch(pkgs: &[PathBuf], pt: &PkgType, dry_run: bool, quiet: bool) -> Result<()> {
    if !is_wsl() {
        display::warning("Not running in WSL — skipping Windows sync.");
        return Ok(());
    }

    if !has_pwsh() && !has_robocopy() {
        display::error("Neither pwsh.exe nor robocopy.exe found.");
        return Ok(());
    }

    for pkg in pkgs {
        let items = prepare_sync_items(pkg, pt);
        if items.is_empty() {
            if !quiet {
                display::info(&format!(
                    "  {}: no files",
                    pkg_basename(pkg),
                ));
            }
            continue;
        }

        if !quiet {
            display::info(&format!(
                "  {}: {} file(s)",
                pkg_basename(pkg),
                items.len(),
            ));
        }

        let counts = sync_batch(&items, dry_run, *SYNC_MAX_CONCURRENT);
        print_sync_summary(&counts, quiet);
    }

    Ok(())
}

pub fn do_stow(pkg: &Path, pt: &PkgType, dry_run: bool) -> Result<()> {
    let target = pt.sync_target().context("no sync target for type")?;
    let pkg_name = pkg
        .file_name()
        .unwrap_or(std::ffi::OsStr::new(""))
        .to_string_lossy();

    if pt.needs_sudo() {
        return stow_file_by_file(pkg, &target, true, dry_run);
    }

    let output = Command::new("stow")
        .args([
            "-v",
            "--adopt",
            "-t",
            &target.to_string_lossy(),
            &pkg_name,
        ])
        .current_dir(&*DOTFILES_DIR)
        .output()?;

    let stderr = String::from_utf8_lossy(&output.stderr);

    if output.status.success() {
        display::success(&format!("Stowed  {}  →  {}", pkg_name, target.display()));
        return Ok(());
    }

    if stderr.contains("existing target is not owned by stow") {
        display::warning(&format!("Stow conflict in {}", pkg_name));
        display::warning("  → Retrying file-by-file (ln -sf) ...");
        return stow_file_by_file(pkg, &target, false, dry_run);
    }

    if stderr.contains("Permission denied") || stderr.contains("cannot stow") {
        display::warning(&format!(
            "Stow permission error in {}: {}",
            pkg_name,
            &stderr[..stderr.len().min(200)]
        ));
    } else {
        return Err(anyhow::anyhow!(
            "Stow failed for {}: {}",
            pkg_name,
            &stderr[..stderr.len().min(300)]
        ));
    }

    Ok(())
}

pub fn stow_file_by_file(pkg: &Path, target: &Path, sudo: bool, dry_run: bool) -> Result<()> {
    for f in list_syncable_files(pkg) {
        let rel = f.strip_prefix(pkg)?;
        let dest = target.join(rel);

        if let Some(parent) = dest.parent() {
            if sudo {
                Command::new("sudo")
                    .args(["mkdir", "-p", &parent.to_string_lossy()])
                    .status()?;
            } else {
                fs::create_dir_all(parent)?;
            }
        }

        let src = f.canonicalize().unwrap_or_else(|_| f.to_path_buf());

        if dry_run {
            display::dim(&format!("      ln -sf {} {}", src.display(), dest.display()));
        } else if sudo {
            Command::new("sudo")
                .args(["ln", "-sf", &src.to_string_lossy(), &dest.to_string_lossy()])
                .status()?;
        } else {
            if dest.exists() || dest.is_symlink() {
                let _ = fs::remove_file(&dest);
            }
            std::os::unix::fs::symlink(&src, &dest)?;
        }
    }

    let pkg_name = pkg
        .file_name()
        .unwrap_or(std::ffi::OsStr::new(""))
        .to_string_lossy();
    display::success(&format!(
        "Linked (file-by-file)  {}  →  {}",
        pkg_name,
        target.display()
    ));

    Ok(())
}

pub fn prepare_sync_items(pkg: &Path, pt: &PkgType) -> Vec<(PathBuf, PathBuf)> {
    list_syncable_files(pkg)
        .into_iter()
        .map(|f| {
            let wp = build_win_path(&f, pkg, pt);
            (f, wp)
        })
        .collect()
}

pub fn sync_batch(
    items: &[(PathBuf, PathBuf)],
    dry_run: bool,
    _max_concurrent: usize,
) -> HashMap<String, usize> {
    let use_robocopy = has_robocopy();

    let total = Arc::new(AtomicUsize::new(0));
    let copied = Arc::new(AtomicUsize::new(0));
    let skipped = Arc::new(AtomicUsize::new(0));
    let missing_source = Arc::new(AtomicUsize::new(0));
    let errors = Arc::new(AtomicUsize::new(0));

    items
        .par_iter()
        .with_max_len(1)
        .try_for_each(|(wsl_path, win_path)| {
            total.fetch_add(1, Ordering::Relaxed);

            if is_excluded(wsl_path) {
                skipped.fetch_add(1, Ordering::Relaxed);
                return Ok::<_, anyhow::Error>(());
            }

            if !wsl_path.exists() {
                missing_source.fetch_add(1, Ordering::Relaxed);
                return Ok(());
            }

            if use_robocopy {
                match robocopy_sync(wsl_path, win_path, dry_run) {
                    Ok(_) => {
                        copied.fetch_add(1, Ordering::Relaxed);
                    }
                    Err(_) => {
                        errors.fetch_add(1, Ordering::Relaxed);
                    }
                }
            } else {
                match pwsh_copy(wsl_path, win_path, wsl_path.is_dir(), dry_run) {
                    Ok(true) => {
                        copied.fetch_add(1, Ordering::Relaxed);
                    }
                    Ok(false) => {
                        errors.fetch_add(1, Ordering::Relaxed);
                    }
                    Err(_) => {
                        errors.fetch_add(1, Ordering::Relaxed);
                    }
                }
            }

            Ok(())
        })
        .ok();

    let mut counts = HashMap::new();
    counts.insert("copied_to_win".to_string(), copied.load(Ordering::Relaxed));
    counts.insert("missing_source".to_string(), missing_source.load(Ordering::Relaxed));
    counts.insert("skipped".to_string(), skipped.load(Ordering::Relaxed));
    counts.insert("error".to_string(), errors.load(Ordering::Relaxed));
    counts
}

fn robocopy_sync(wsl_src: &Path, win_dst: &Path, dry_run: bool) -> Result<()> {
    let wsl_unc = format!(
        "\\\\wsl$\\{}\\{}",
        *WSL_DISTRO_NAME,
        wsl_src.to_string_lossy().replace('/', "\\")
    );

    let win_str = format!(
        "C:\\{}",
        win_dst
            .strip_prefix(&*MNT_C)
            .unwrap_or(win_dst)
            .to_string_lossy()
            .replace('/', "\\")
    );

    if dry_run {
        display::dim(&format!(
            "    DRY-RUN  robocopy {} {} /MIR /MT:8",
            wsl_src.display(),
            win_str,
        ));
        return Ok(());
    }

    let mut cmd = if wsl_src.is_dir() {
        let mut c = Command::new("robocopy.exe");
        c.args([
            &wsl_unc,
            &win_str,
            "/MIR",
            "/MT:8",
            "/R:1",
            "/W:1",
            "/NJH",
            "/NJS",
            "/NP",
            "/XF",
            ".git",
        ]);
        c
    } else {
        let parent = wsl_unc
            .rsplit_once('\\')
            .map(|(p, _)| p.to_string())
            .unwrap_or(wsl_unc.clone());
        let file = wsl_src
            .file_name()
            .unwrap_or(std::ffi::OsStr::new(""))
            .to_string_lossy()
            .to_string();
        let mut c = Command::new("robocopy.exe");
        c.args([
            &parent,
            &win_str,
            &file,
            "/R:1",
            "/W:1",
            "/NJH",
            "/NJS",
            "/NP",
        ]);
        c
    };

    let status = cmd
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()?;

    let code = status.code().unwrap_or(8);
    if code < 8 {
        Ok(())
    } else {
        bail!("robocopy failed with exit code {}", code)
    }
}

fn pwsh_copy(wsl_src: &Path, win_dst: &Path, is_dir: bool, dry_run: bool) -> Result<bool> {
    let wsl_unc = format!(
        "\\\\wsl$\\{}\\{}",
        *WSL_DISTRO_NAME,
        wsl_src.to_string_lossy().replace('/', "\\")
    );

    let win_str = format!(
        "C:\\{}",
        win_dst
            .strip_prefix(&*MNT_C)
            .unwrap_or(win_dst)
            .to_string_lossy()
            .replace('/', "\\")
    );

    let copy_cmd = if is_dir {
        format!("xcopy /E /I /Y \"{}\" \"{}\"", wsl_unc, win_str)
    } else {
        format!("copy /Y \"{}\" \"{}\"", wsl_unc, win_str)
    };

    if dry_run {
        display::dim(&format!(
            "    DRY-RUN  pwsh.exe -Command cmd /c {}",
            copy_cmd
        ));
        return Ok(true);
    }

    let output = Command::new("pwsh.exe")
        .args(["-NoProfile", "-Command", &format!("cmd /c {}", copy_cmd)])
        .output()?;

    if output.status.success() {
        Ok(true)
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr);
        display::error(&format!("Windows copy failed: {}", &stderr[..stderr.len().min(200)]));
        Ok(false)
    }
}

fn print_sync_summary(counts: &HashMap<String, usize>, quiet: bool) {
    if quiet {
        return;
    }

    let mut parts: Vec<String> = Vec::new();
    if let Some(&c) = counts.get("copied_to_win")
        && c > 0 {
            parts.push(format!("{} copied", c));
        }
    if let Some(&s) = counts.get("skipped")
        && s > 0 {
            parts.push(format!("{} skipped", s));
        }
    if let Some(&m) = counts.get("missing_source")
        && m > 0 {
            parts.push(format!("{} missing", m));
        }
    if let Some(&e) = counts.get("error")
        && e > 0 {
            parts.push(format!("{} errors", e));
        }

    if !parts.is_empty() {
        display::info(&format!("    Result: {}", parts.join(", ")));
    }
}
